src/server/Ebenezer/GameDefine.h

```diff
@@ -543,6 +543,7 @@ struct _WARP_INFO {
     float fY;
     float fZ;
     float fR;
+    short sNation;
```

Comment:
Why add this if it isn't being used anywhere?

Reply:
Because without it loading any map will give screwed up warps.
Unless you want to stick with the oldest available maps elmorad and karus zone, they do work out of the box.
But before Moradon and KE was introduced this was added, it's a matter of preference I can either go with map.cpp load warplist readfile and do a skip size of short. Or we can just adjust the size of warp struct to fit newer SMD's.
See screenshot I included in the opening post. I remember twostars saying that the SMD's were horribly wrong/damaged/corrupted. Nobody ever took this much time as I dissected SMD's with hex editor.
I will publish the smd files along with this code once I'm done.

Reply:
If you want to diagnose your own files, you can use my provided sample code this goes straight into map.cpp ebenezer:
Now you can export all warps to a text file, adjust struct of warpinfo with a snation and try without it, if you got mixed smd files you will see mixed results some zones proper warps and some with invalid numbers because of the missing short.

void C3DMap::LoadWarpList(HANDLE hFile) {
    int          WarpCount = 0;
    DWORD        dwNum;
    _WARP_INFO * pWarp = NULL;

    ReadFile(hFile, &WarpCount, 4, &dwNum, NULL);

    for (int i = 0; i < WarpCount; i++) {
        pWarp = new _WARP_INFO;

        ReadFile(hFile, pWarp, sizeof(_WARP_INFO), &dwNum, NULL);

        if (!m_WarpArray.PutData(pWarp->sWarpID, pWarp)) {
            TRACE("Warp list PutData Fail - %d\n", pWarp->sWarpID);
            delete pWarp;
            pWarp = NULL;
        }
        UpdateWarpInfo(pWarp);
    }
}

struct Warp {
    short sWarpID;
    char  strWarpName[32];
    char  strAnnounce[256];
    DWORD dwPay;
    short sZone;
    float fX;
    float fY;
    float fZ;
    float fR;
    short sNation;
};

void C3DMap::UpdateWarpInfo(_WARP_INFO * pWarp) {
    if (!pWarp) {
        return;
    }
    CStdioFile      file;
    CString         line;
    CArray<CString> lines;

    if (file.Open(_T("warpinfo.txt"), CFile::modeReadWrite | CFile::typeText)) {
        while (file.ReadString(line)) {
            line.TrimRight(); // Remove trailing whitespace
            if (line.IsEmpty()) {
                continue; // Skip empty lines
            }

            Warp    warp;
            CString remaining;
            if (sscanf_s(line, "%hd\t%31[^\t]\t%255[^\t]\t%d\t%hd\t%f\t%f\t%f\t%f\t%d", &warp.sWarpID, warp.strWarpName,
                         32, warp.strAnnounce, 256, &warp.dwPay, &warp.sZone, &warp.fX, &warp.fY, &warp.fZ, &warp.fR,
                         &warp.sNation) == 10) {
                if (warp.sWarpID == pWarp->sWarpID) {
                    // Load the data from the file
                    strncpy_s(pWarp->strWarpName, warp.strWarpName, sizeof(pWarp->strWarpName));
                    strncpy_s(pWarp->strAnnounce, warp.strAnnounce, sizeof(pWarp->strAnnounce));
                    pWarp->dwPay = warp.dwPay;
                    pWarp->sZone = warp.sZone;
                    pWarp->fX = warp.fX;
                    pWarp->fY = warp.fY;
                    pWarp->fZ = warp.fZ;
                    pWarp->fR = warp.fR;
                    pWarp->sNation = warp.sNation;

                    file.Close();
                    return; // Found matching "warpid" in the file, loaded the data
                }
            }
            lines.Add(line); // Store the current line for rewriting
        }

        // Matching "warpid" not found, insert a new line with the provided pWarp data
        CString newLine;
        newLine.Format(_T("%hd\t%s\t%s\t%d\t%hd\t%f\t%f\t%f\t%f\t%d"), pWarp->sWarpID, pWarp->strWarpName,
                       pWarp->strAnnounce, pWarp->dwPay, pWarp->sZone, pWarp->fX, pWarp->fY, pWarp->fZ, pWarp->fR,
                       pWarp->sNation);
        lines.Add(newLine);

        // Rewrite the file with the updated lines
        file.SeekToBegin();
        file.SetLength(0); // Clear the file
        for (int i = 0; i < lines.GetSize(); i++) {
            file.WriteString(lines[i]);
            file.WriteString(_T("\n")); // Add a new line character
        }

        file.Close();
    } else {
        // File does not exist, create it and insert the pWarp data
        if (file.Open(_T("warpinfo.txt"), CFile::modeCreate | CFile::modeWrite | CFile::typeText)) {
            CString newLine;
            newLine.Format(_T("%hd\t%s\t%s\t%d\t%hd\t%f\t%f\t%f\t%f\t%d"), pWarp->sWarpID, pWarp->strWarpName,
                           pWarp->strAnnounce, pWarp->dwPay, pWarp->sZone, pWarp->fX, pWarp->fY, pWarp->fZ, pWarp->fR,
                           pWarp->sNation);
            file.WriteString(newLine);
            file.WriteString(_T("\n")); // Add a new line character

            file.Close();
        } else {
            AfxMessageBox(_T("Failed to create warpinfo.txt file."));
        }
    }
}

Reply:
Just to add some context and another perspective, some gate objects are nation 0 which means for both, but that doesn't mean you will get Karus and Elmorad warp entries from the same object, that differs if you are invading Karus you are not allowed to teleport to Karus entries but you are allowed to use the object to warp back. (hypothesis) I'm not even going to confirm my suspicion or hypothesis, sooner or later this will have to be done and for now it saves me a lot of work. And if we go to KE SMD's they will have the extra short anyway >.>

Reply:
I am already in the process for doing that I just need gren light and approval then I will upload them and hopefully merge.
<3

Reply:
> I imagine this would prevent our current maps from working, right? Could we update our old maps first by adding a short in the right place? This way we can still load our updated older maps.
![image](https://github.com/ko4life-net/ko/assets/10597986/22f10356-2e6f-4df2-8f06-c5ecc97a3632)

Tested and they work fine so far.
We will need to add event zones for eslant tiles dbo.event is empty.
Also we need to adjust client zones.tbl and zones directory, the 1097 Colony Zone is just way too outdated......
Let's do the rest in different commits okay? (Loading 1264 maps (int version, int stringlength, readsizeofchar + add null termination to terrain loading and opd loading).
I don't want to clutter this Draft too much..

DB table zone_info:
<HTML>
<BODY>
<!--StartFragment--><TABLE><TR><TD>1</TD><TD>2</TD><TD>elmorad_1212.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>163700</TD><TD>41800</TD><TD>38500</TD><TD>1</TD><TD>0</TD></TR><TR><TD>1</TD><TD>1</TD><TD>karus_1212.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>37120</TD><TD>157300</TD><TD>38500</TD><TD>1</TD><TD>0</TD></TR><TR><TD>1</TD><TD>101</TD><TD>battle_0810.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>23700</TD><TD>14000</TD><TD>38500</TD><TD>3</TD><TD>0</TD></TR><TR><TD>1</TD><TD>102</TD><TD>battle_0810.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>23700</TD><TD>14000</TD><TD>38500</TD><TD>3</TD><TD>0</TD></TR><TR><TD>1</TD><TD>201</TD><TD>freezone_1217.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>1000</TD><TD>1000</TD><TD>0</TD><TD>2</TD><TD>0</TD></TR><TR><TD>1</TD><TD>21</TD><TD>moradon_1021.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>31200</TD><TD>40200</TD><TD>0</TD><TD>1</TD><TD>0</TD></TR><TR><TD>1</TD><TD>11</TD><TD>k_eslant.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>15000</TD><TD>15000</TD><TD>15000</TD><TD>0</TD><TD>0</TD></TR><TR><TD>1</TD><TD>12</TD><TD>e_eslant.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>15000</TD><TD>15000</TD><TD>15000</TD><TD>0</TD><TD>0</TD></TR><TR><TD>1</TD><TD>30</TD><TD>siege_1110.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>1000</TD><TD>1000</TD><TD>0</TD><TD>2</TD><TD>0</TD></TR><TR><TD>1</TD><TD>32</TD><TD>abyssb_0925.smd &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD><TD>1000</TD><TD>1000</TD><TD>0</TD><TD>2</TD><TD>0</TD></TR></TABLE>
<!--EndFragment-->
</BODY>
</HTML>

[servermaps.zip](https://github.com/ko4life-net/ko/files/12171907/servermaps.zip)

Reply:
Agree with @srmeier here.